services/tracing/tracing/tracing.go

```diff
@@ -1,9 +1,135 @@
 package tracing
 
+import (
+	"time"
+
+	"github.com/samber/lo"
+)
+
+type Status int
+
+const (
+	Unknown Status = iota
+	Canceled
+	Pending
+	Success
+	InProgress
+	Failed
+)
+
+func (s Status) String() string {
+	switch s {
+	case Unknown:
+		return "unknown"
+	case Canceled:
+		return "canceled"
+	case Pending:
+		return "pending"
+	case Success:
+		return "success"
+	case InProgress:
+		return "in progress"
+	case Failed:
+		return "failed"
+	}
+	return "unknown"
+}
+
 type Step struct {
-	TracingID      string `db:"tracing_id"`
-	StepIndex      int64  `db:"step_index"`
-	StepsRemaining int64  `db:"steps_remaining"`
-	StartTime      int64  `db:"start_time"`
-	Error          string `db:"error"`
+	TracingID      string
+	StepIndex      int64
+	StepsRemaining int64
+	StartTime      int64
+	DeviceId       *int64
+	Error          *string
+}
+
+// EnrichedStep contains extra properties derived from the data stored in the database using the Step model
+type EnrichedStep struct {
+	Step
+	HighestCollectiveStatus Status
+	Status                  Status
+	Duration                time.Duration
+}
+
+type EnrichedSteps []EnrichedStep
+
+func (es EnrichedSteps) TotalStartTime() int64 {
+	return lo.MinBy(es, func(item, min EnrichedStep) bool {
+		// A StartTime with value 0 is considered to be not set
+		return item.StartTime > 0 && item.StartTime < min.StartTime
+	}).StartTime
+}
+
+func (es EnrichedSteps) TotalStatus() Status {
+	if len(es) == 0 {
+		return Unknown
+	}
+	return es[0].HighestCollectiveStatus
+}
+
+func (es EnrichedSteps) DeviceId() *int64 {
+	if val, ok := lo.Find(es, func(item EnrichedStep) bool {
+		return item.DeviceId != nil
+	}); ok {
+		return val.DeviceId
+	}
+	return nil
+}
+
+func (es EnrichedSteps) AllSteps() EnrichedSteps {
+	if len(es) == 0 {
+		return []EnrichedStep{}
+	}
+
+	// Enriched Steps are only models from the database, some steps do not exist yet in the database, however we do
+	// want to include them in the all steps list. Using the status from the last step we can derive the state of all remaining steps
+
+	// The last step status and steps remaining determines if any steps need to be added to the step list
+	lastStep := es[len(es)-1]
+	remainingStatus := Pending
+	if lastStep.Status == Failed {
+		// When the last step in the list has failed all the remaining steps are canceled
+		remainingStatus = Canceled
+	}
+
+	steps := lo.Times(len(es)+int(lastStep.StepsRemaining), func(index int) EnrichedStep {
```

Comment:
Does this take into consideration that Step 1 could possible not exist while step2 does?

Reply:
good point! it doesn't. I updated the code to add a step with status 'unknown' should a further step arrive first. Is this desired behaviour or should we automatically set the status to succesful since a next step record can only exist if the step before it was OK?

Reply:
Should now be succes as the next step had already been reached. This has been discussed IRL